packages/backend/src/models/ProjectModel/ProjectModel.ts

```diff
@@ -2082,15 +2083,59 @@ export class ProjectModel {
         return { name };
     }
 
+    private static isSemanticLayerConnectionValid(
+        semanticLayerConnection: SemanticLayerConnectionUpdate,
+    ): boolean {
+        const { type } = semanticLayerConnection;
+        switch (type) {
+            case SemanticLayerType.DBT:
+                return Boolean(
+                    semanticLayerConnection.domain &&
+                        semanticLayerConnection.environmentId &&
+                        semanticLayerConnection.token,
+                );
+            case SemanticLayerType.CUBE:
+                return Boolean(
+                    semanticLayerConnection.domain &&
+                        semanticLayerConnection.token,
+                );
+            default:
+                return assertUnreachable(
+                    type,
+                    `Unknown semantic layer connection type: ${type}`,
+                );
+        }
+    }
+
     async updateSemanticLayerConnection(
         projectUuid: string,
-        connection: SemanticLayerConnection | undefined,
+        connectionUpdate: SemanticLayerConnectionUpdate,
     ) {
+        const { semanticLayerConnection: currentSemanticLayerConnection } =
+            await this.getWithSensitiveFields(projectUuid);
+
+        // ? Merging so the partial update only overwrites the existing properties, even when the current connection is undefined since merge will take ALL properties
+        const shouldMerge =
+            !currentSemanticLayerConnection ||
+            connectionUpdate.type === currentSemanticLayerConnection.type;
+
+        const updatedSemanticLayerConnection = shouldMerge
+            ? merge(currentSemanticLayerConnection, connectionUpdate)
+            : connectionUpdate;
+
+        if (
+            !ProjectModel.isSemanticLayerConnectionValid(
+                updatedSemanticLayerConnection,
+            )
+        ) {
+            throw new ParameterError('Invalid semantic layer connection');
+        }
```

Comment:
It should only merge the current connection with the partial update when:
- There is no current connection since it will take all properties from the partial
- When type is the same, this is so that we don't end up with rogue properties being saved to the wrong type

Before saving we always check if the `updatedSemanticLayerConnection` is valid (has all properties needed for the semantic layer type), this is to ensure that we never create a connection with missing properties

Reply:
makes sense. nice work!